modules/storages/lib/api/v3/file_links/file_link_representer.rb

```diff
@@ -0,0 +1,136 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2022 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module API
+  module V3
+    module FileLinks
+      class FileLinkRepresenter < ::API::Decorators::Single
+        include API::Decorators::LinkedResource
+        include API::Decorators::DateProperty
+        include ::API::Caching::CachedRepresenter
+
+        property :id
+
+        date_time_property :created_at, writeable: false
+
+        date_time_property :updated_at, writeable: false
+
+        property :originData,
+                 exec_context: :decorator,
+                 getter: ->(*) { make_origin_data(represented) },
+                 setter: ->(fragment:, **) do
+                   parse_origin_data(fragment).each do |attribute, value|
+                     represented[attribute] = value
+                   end
+                 end
+
+        link :self do
+          {
+            href: api_v3_paths.file_link(represented.id)
+          }
+        end
+
+        link :delete, cache_if: -> { user_allowed_to_manage?(represented) } do
+          {
+            href: api_v3_paths.file_link(represented.id),
+            method: :delete
+          }
+        end
+
+        link :creator do
+          {
+            href: api_v3_paths.user(represented.creator_id),
+            title: represented.creator.name
+          }
+        end
+
+        link :originOpen do
+          {
+            href: ::Storages::StorageUrlService.new(represented).call('open').result
```

Comment:
This is a magic string again. I'd argue that for the time being, the parameter is not necessary at all. If it is, use a constant.

Reply:
What is a magic string exactly? `'open'`?

Reply:
Usually I would use an enum for this. When I checked particular other places, I saw those "frozen" list of strings that were used as constants.

I'd gladly put it to a constant and remove the magic string. But remove it completely might just result in readding it in 2 weeks. The URL service is meant to be initialized on a file link object, then called upon for the different use cases to return a storage based url. The use cases are `open` and `download` for the time being. (in future there might be a third, `open-edit`)

Reply:
> What is a magic string exactly? 'open'?

Yes, that 'open' is what I was referring to. 

Having taken another look at the `Storages::StorageUrlService`. Maybe it should not be a service since it does not do/process anything. It is more akin to a helper. How about we turn it into a module that can then be mixed in where it is used. That way, we can simply add methods to it:

```
module StorageUrlHelper
  def storage_url_open(file_link)
    "#{file_link.storage.host}/f/#{file_link.origin_id}"
  end

  def storage_url_download(file_link)
    ..
  end
end

# when using

class SomeClass
  include StorageUrlHelper

  ...
     href: storage_url_open(file_link)
end
```

Reply:
Done in 9e968c4752.

It simplifies the code as it does not account for the `download` case yet. That's ok. Abstractions are harder to build when they are speculated. We should stick with the simplest implementation that works until we introduce `download` and then revisit.